tests/regression/src/lib.rs

```diff
@@ -87,70 +146,164 @@ mod tests {
         .unwrap();
     }
 
-    /// Test which creates a TestPair from config using `rsa_4096_sha512`. Only measures a pair handshake. 
+    /// Test which creates a TestPair from config using `rsa_4096_sha512`. Only measures a pair handshake.
     #[test]
     fn test_rsa_handshake() {
         valgrind_test("test_rsa_handshake", |ctrl| {
             ctrl.stop_instrumentation();
-            // Example usage with RSA keypair (default)
             let keypair_rsa = CertKeyPair::default();
             let config = set_config(&security::DEFAULT_TLS13, keypair_rsa)?;
-            // Create a pair (client + server) using that config, start handshake measurement
             let mut pair = TestPair::from_config(&config);
-            // Assert a successful handshake
             ctrl.start_instrumentation();
             assert!(pair.handshake().is_ok());
             ctrl.stop_instrumentation();
             Ok(())
         })
         .unwrap();
     }
-    /// Function to run specified test using valgrind
+
+    /// Function to run specified test with valgrind for performance profiling.
     fn run_valgrind_test(test_name: &str) {
         let exe_path = std::env::args().next().unwrap();
-        create_dir_all(Path::new("target/cg_artifacts")).unwrap();
-        let output_file = format!("target/cg_artifacts/cachegrind_{}.out", test_name);
-        let output_command = format!("--cachegrind-out-file={}", &output_file);
-        let mut command = Command::new("valgrind");
-        command
-            .args(["--tool=cachegrind", &output_command, &exe_path, test_name])
-            // Ensures that the recursive call is made to the actual harness code block rather than back to this function
-            .env_remove("ENABLE_VALGRIND"); 
+        let commit_hash = git::get_current_commit_hash();
+        let output_file = create_raw_profile_path(test_name, &commit_hash);
+        let command = build_valgrind_command(&exe_path, test_name, &output_file);
 
         println!("Running command: {:?}", command);
-        let status = command.status().expect("Failed to execute valgrind");
+        execute_command(command);
+
+        let annotate_output = run_annotation(&output_file);
```

Comment:
I think moving to generic filepaths + write made things neater, but I wonder if we could reduce complexity even more by representing the Raw/Annotated/Diff profiles as structs/type?
```
RawProfile::new(test_name, commit_it);
let annotated_profile = raw_profile.annotate();
let diff_profiles = DiffProfile::from(annotated_new, annotated_old);
```

Right now a lot of the complexity exists in free functions which isn't necessarily bad, but there's a large number of them which makes it a bit hard for me to follow.